File: 2024/Day 15/Python/main.py
"""Advent of Code - 2024 - Day 15"""

type Vec = tuple[int, int]

def add(a: Vec, b: Vec):
    """adds 2 Vecs together"""
    return (a[0] + b[0], a[1] + b[1])

def move_box1(box_pos: Vec, d: Vec, boxes: set[Vec], walls: set[Vec]):
    """moves a box using part 1 rules"""
    next_pos = add(box_pos, d)
    if next_pos in walls or (next_pos in boxes and not move_box1(next_pos, d, boxes, walls)):
        return False
    boxes.add(next_pos)
    boxes.remove(box_pos)
    return True

def move_box2(box_pos: Vec, d: Vec, boxes: set[Vec], walls: set[Vec]):
    """moves a box using part 2 rules"""
    next_pos = add(box_pos, d)

    if d == (1, 0): # right
        if add(box_pos, (2, 0)) in walls:
            return False
        if add(box_pos, (2, 0)) in boxes:
            if move_box2(add(box_pos, (2, 0)), d, boxes, walls):
                boxes.add(next_pos)
                boxes.remove(box_pos)
                return True
            return False
        boxes.add(next_pos)
        boxes.remove(box_pos)
        return True
    if d == (-1, 0): # left
        if next_pos in walls:
            return False
        if add(box_pos, (-2, 0)) in boxes:
            if move_box2(add(box_pos, (-2, 0)), d, boxes, walls):
                boxes.add(next_pos)
                boxes.remove(box_pos)
                return True
            return False
        boxes.add(next_pos)
        boxes.remove(box_pos)
        return True
    # up or down
    if next_pos in walls or add(next_pos, (1, 0)) in walls:
        return False

    il = check_intersect(next_pos, boxes)
    ir = check_intersect(add(next_pos, (1, 0)), boxes)
    if il and ir:
        if il != ir:
            copy_boxes = boxes.copy()
            if move_box2(il, d, copy_boxes, walls) and move_box2(ir, d, copy_boxes, walls):
                boxes.clear()
                boxes |= copy_boxes
                boxes.add(next_pos)
                boxes.remove(box_pos)
                return True
            return False
        if move_box2(il, d, boxes, walls):
            boxes.add(next_pos)
            boxes.remove(box_pos)
            return True
        return False
    if il:
        if move_box2(il, d, boxes, walls):
            boxes.add(next_pos)
            boxes.remove(box_pos)
            return True
        return False
    if ir:
        if move_box2(ir, d, boxes, walls):
            boxes.add(next_pos)
            boxes.remove(box_pos)
            return True
        return False
    boxes.add(next_pos)
    boxes.remove(box_pos)
    return True

def check_intersect(pos, boxes):
    """checks if pos intersects with a box"""
    if pos in boxes:
        return pos
    if add(pos, (-1, 0)) in boxes:
        return add(pos, (-1, 0))
    return None

with open(0, encoding="utf-8") as f:
    warehouse, moves = f.read().split("\n\n")
moves: str = "".join(moves.splitlines())
warehouse = warehouse.splitlines()
w = len(warehouse[0])
h = len(warehouse)


walls1 = set()
boxes1 = set()
robot1 = (-1, -1)
walls2 = set()
boxes2 = set()
robot2 = (-1, -1)

for y, line in enumerate(warehouse):
    for x, c in enumerate(line):
        if c == '#':
            walls1.add((x, y))
            walls2.add((2*x, y))
            walls2.add((2*x+1, y))
        elif c == 'O':
            boxes1.add((x, y))
            boxes2.add((2*x, y))
        elif c == '@':
            robot1 = (x, y)
            robot2 = (2*x, y)

dirs = {'^': (0, -1), 'v': (0, 1), '<': (-1, 0), '>': (1, 0)}
for move in moves:
    direction = dirs[move]
    npos = (robot1[0] + direction[0], robot1[1] + direction[1])
    if npos in boxes1:
        if move_box1(npos, direction, boxes1, walls1):
            robot1 = npos
    elif npos not in walls1:
        robot1 = npos

print(sum(100*pos[1] + pos[0] for pos in boxes1))

for move in moves:
    direction = dirs[move]
    npos = add(robot2, direction)
    intersect = check_intersect(npos, boxes2)
    if intersect:
        if move_box2(intersect, direction, boxes2, walls2):
            robot2 = npos
    elif npos not in walls2:
        robot2 = npos

print(sum(100*pos[1] + pos[0] for pos in boxes2))

File: 2024/Day 15/C++/main.cpp
#include <string>
#include <iostream>
#include <vector>
#include <map>
#include <algorithm>

template <typename T, typename U>
std::pair<T, U> operator+(const std::pair<T, U> &l, const std::pair<T, U> &r)
{
    return {l.first + r.first, l.second + r.second};
}

bool has(std::vector<std::pair<int, int>> &vec, std::pair<int, int> &val)
{
    return std::find(vec.begin(), vec.end(), val) != vec.end();
}

bool move_box1(std::pair<int, int> &box_pos, std::pair<int, int> &dir, std::vector<std::pair<int, int>> &boxes, std::vector<std::pair<int, int>> &walls)
{
    std::pair<int, int> npos = box_pos + dir;
    if (has(walls, npos) || (has(boxes, npos) && !move_box1(npos, dir, boxes, walls)))
    {
        return false;
    }
    std::vector<std::pair<int, int>>::iterator it = std::find(boxes.begin(), boxes.end(), box_pos);
    if (it != boxes.end())
    {
        *it = npos;
    }
    else
    {
        std::cout << "error\n";
    }
    return true;
}

int main(int argc, char *argv[])
{
    std::string line;
    bool isMaze = true;
    std::string moves;
    int y = 0;
    std::vector<std::pair<int, int>> walls1;
    std::vector<std::pair<int, int>> boxes1;
    std::pair<int, int> robot1;
    while (std::getline(std::cin, line))
    {
        if (line == "")
        {
            isMaze = false;
        }
        else if (isMaze)
        {
            for (int x = 0; x < line.length(); x++)
            {
                switch (line[x])
                {
                case '#':
                    walls1.push_back({x, y});
                    break;
                case 'O':
                    boxes1.push_back({x, y});
                    break;
                case '@':
                    robot1 = {x, y};
                    break;
                default:
                    break;
                }
            }
            y++;
        }
        else
        {
            moves += line;
        }
    }

    const std::map<char, std::pair<int, int>> dirs = {{'^', {0, -1}}, {'v', {0, 1}}, {'<', {-1, 0}}, {'>', {1, 0}}};

    for (char move : moves)
    {
        std::pair<int, int> dir = dirs.at(move);
        std::pair<int, int> npos = robot1 + dir;
        if (has(boxes1, npos))
        {
            if (move_box1(npos, dir, boxes1, walls1))
            {
                robot1 = npos;
            }
        }
        else if (!has(walls1, npos))
        {
            robot1 = npos;
        }
    }

    int score1 = 0;
    for (std::pair<int, int> pos : boxes1)
    {
        score1 += pos.first + 100 * pos.second;
    }
    std::cout << score1 << "\n";
}